### PR TITLE
[Feature] `OrderedDict.move_to_end`

### LINKED ISSUE #4408: Add OrderedDict.move_to_end
API doc is here: https://docs.python.org/3/library/collections.html#collections.OrderedDict.move_to_end

Code would be added to: https://github.com/adafruit/circuitpython/blob/main/py/objdict.c#L563 It'll need a separate locals so that move_to_end isn't added for regular dictionaries.

This will fix https://github.com/adafruit/Adafruit_CircuitPython_BLE_BroadcastNet/issues/18

### PR DESCRIPTION
**Pull request by elpekenin**
Closes #4408

Adds the method in the title. First time i write a C-Python binding, please let me know if anything should be changed
- Should maybe remove the doctstring to follow other funcs not having it on the file
- Is there a way to extend existing array instead of duplicating? Could save a bit of space

---
My little test:
```py
Adafruit CircuitPython 8.2.0-47-g8050b07a6-dirty on 2023-07-30; WeAct ESP32S3-B-N16R8 with ESP32S3R8

>>> from collections import OrderedDict
>>> d = OrderedDict.fromkeys("abcd")
>>> print(d)
OrderedDict({'a': None, 'b': None, 'c': None, 'd': None})
>>> d.move_to_end("a")
>>> print(d)
OrderedDict({'b': None, 'c': None, 'd': None, 'a': None})
>>> d.move_to_end("d", False)
>>> print(d)
OrderedDict({'d': None, 'b': None, 'c': None, 'a': None})
>>> d.move_to_end("b", last=False)
>>> print(d)
OrderedDict({'b': None, 'd': None, 'c': None, 'a': None})
```

### CODE REVIEW

**Comment by jepler** [comment]
Unfortunately, this change is overflowing the available flash size on a few builds :-/ there's no easy solution for this, the resolution may be to choose not to include it, or to disable some other feature on the failing board.

**Comment by elpekenin** [comment]
Yeah, should i add a new #define + #ifdef to control whether the API's code gets added?

Could also add the method to dict's table, but this breaks worse than fixes:
- i dont think the method exists on regular dicts
- would only save a few bytes (not duplicating a couple of pointers and whatnot across two tables)

**Comment by dhalbert** [comment]
> Yeah, should i add a new #define + #ifdef to control whether the API's code gets added?

Yup. We use `#if` instead of `#ifdef` so we make sure that the setting has a defined value. There are many examples of that in various places.

Instead of making yet another flag, there is a kitchen-sink one called `MICROPY_CPYTHON_COMPAT` which turns on various extra features. I think this feature should be controlled by that for now, so that it's not added for the smallest builds. In the long run I wanted to refactor that flag into separate flags, but in the short run that would be an easy flag to use.

However, the boards that are overflowing I don't think would be fixed by that. We need to look at the SAMD51 boards that are close to overflowing and come up with a general strategy to shrink them. @jepler started to do this for some other boards. For instance, we may be able to shrink `ulab` significantly. So let's hold this PR up for a bit while we look jharder. Since this would be a 9.0.0 feature, merging is not urgent.

**Comment by elpekenin** [comment]
No worries, we're not in a rush :D

Will add the new `#if` soon, and make the `#else` do `ordered_dict_locals_dict = dict_locals_dict`

Let me know if any code should be re-written for readability/matching codestyle of the repo

**Comment by elpekenin** [comment]
Just in case: Dont merge yet, i think key is positional-only right now (forgot its QSTR on the structure which parses args)

File: py/objdict.c [hunk 1]
```diff
@@ -643,7 +644,7 @@ STATIC mp_obj_t dict_move_to_end(size_t n_args, const mp_obj_t *pos_args, mp_map
     enum { ARG_self, ARG_key, ARG_last };
     static const mp_arg_t allowed_args[] = {
         { MP_QSTR_, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE } },
```
Comment (dhalbert) [inline py/objdict.c:646]: This is not generally how we do `self` handling. Instead, delete this line, and use the suggestions below.
```suggestion
```

File: py/objdict.c [hunk 2]
```diff
@@ -643,7 +644,7 @@ STATIC mp_obj_t dict_move_to_end(size_t n_args, const mp_obj_t *pos_args, mp_map
     enum { ARG_self, ARG_key, ARG_last };
```
Comment (dhalbert) [inline py/objdict.c:644]: Don't use `ARG_self`:
```suggestion
    enum { ARG_key, ARG_last };
```

File: py/objdict.c (hunk 1 above)
Comment (dhalbert) [reply py/objdict.c:646]: Change the `mp_arg_parse_all()` to:
```c
    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
```
Change the `self` assignment to:
```c
    mp_obj_dict_t *self = MP_OBJ_TO_PTR(pos_args[0]);
```

**Review by dhalbert — CHANGES_REQUESTED** [review]

**Comment by dhalbert** [comment]
Here are the test failure details:
```
 True
 abcdefghij
 abcdefghij
-odict_keys(['a', 'b'])
-odict_keys(['b', 'a'])
-odict_keys(['a', 'b'])
-odict_keys(['b', 'a'])
-odict_keys(['a', 'b'])
-odict_keys(['b', 'a'])
-odict_keys(['a', 'b'])
+dict_keys(['a', 'b'])
+dict_keys(['b', 'a'])
+dict_keys(['a', 'b'])
+dict_keys(['b', 'a'])
+dict_keys(['a', 'b'])
+dict_keys(['b', 'a'])
+dict_keys(['a', 'b'])

FAILURE /home/runner/work/circuitpython/circuitpython/tests/results/basics_ordereddict1.py
```
Looks like it changed from printing `odict_keys` to `dict_keys`.

See https://github.com/adafruit/circuitpython/actions/runs/5744164923/job/15569982403?pr=8234, and then click "Print failure info" to see the details I pasted above.

**Comment by elpekenin** [comment]
Should i have that many tests, for every combination of positional and keyword usage (made that to confirm i made things correctly), or would it be better to just check moving either way?

**Comment by dhalbert** [comment]
> Should i have that many tests, for every combination of positional and keyword usage (made that to confirm i made things correctly), or would it be better to just check moving either way?

I think the main thing is to test the moving and edge cases (e.g. on an empty dict, key not found, etc.) not so much the arg parsing. It doesn't hurt but the testing is more about the functionality than the arg parsing. Also test arg validation maybe: suppose you pass a non-bool for the bool arg.

**Comment by elpekenin** [comment]
Tests are now checking (all?) edge cases + are successful, however flash is still too large...

**Comment by jepler** [comment]
I wrote #8258 which also implements move_to_end, but at smaller flash cost. Thank you for the original implementation!

**Comment by dhalbert** [comment]
Closed in favor of #8258.
